File: fedex-main/globex-fedex-main/frontend/src/app/core/services/help-center.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import { catchError } from 'rxjs/operators';

import { API_BASE_URL } from '../api.config';
import { FALLBACK_ARTICLE_SUMMARIES, fallbackArticle } from './help-articles.fallback';

export interface HelpResource {
  id: string;
  title: string;
  description: string;
  articlesCount: number;
  readTime: string;
  category: string;
}

export interface HelpMetrics {
  averageResponse: string;
  resolutionRate: string;
  ticketsSolved: number;
  satisfaction: string;
  supportOnline: boolean;
}

export interface HelpContactOption {
  id: string;
  title: string;
  description: string;
  status: string;
  cta: string;
  action: string;
}

export interface HelpArticleStep {
  title: string;
  body: string;
}

export interface HelpArticleFaqItem {
  question: string;
  answer: string;
}

export interface HelpArticleSection {
  type: string;
  heading?: string | null;
  body?: string | null;
  items?: string[] | null;
  steps?: HelpArticleStep[] | null;
  faq?: HelpArticleFaqItem[] | null;
}

export interface HelpArticleSummary {
  id: string;
  slug: string;
  title: string;
  category: string;
  summary: string;
  readTime: string;
  updatedAt: string;
}

export interface HelpArticle extends HelpArticleSummary {
  sections: HelpArticleSection[];
  relatedSlugs: string[];
}

export interface CreateTicketPayload {
  subject: string;
  category: string;
  priority: string;
  message: string;
  attachmentUrl?: string | null;
}

export interface CreateTicketResponse {
  success: boolean;
  ticketId: string;
  id: number;
  status: string;
  createdAt: string;
}

export interface SupportTicketSummary {
  id: number;
  ticket_number?: string;
  subject: string;
  message: string;
  category: string;
  priority: string;
  status: string;
  attachment_url?: string | null;
  created_at: string;
  updated_at?: string | null;
}

export interface SupportTicketListResponse {
  items: SupportTicketSummary[];
}

const FALLBACK_RESOURCES: HelpResource[] = [
  {
    id: 'tracking',
    title: 'Tracking issues',
    description: 'Track parcels, ETA, exceptions and delivery problems.',
    articlesCount: 8,
    readTime: '2 min read',
    category: 'tracking',
  },
  {
    id: 'documents',
    title: 'Documents & POD',
    description: 'Manage proofs of delivery, exports and archived reports.',
    articlesCount: 6,
    readTime: '3 min read',
    category: 'documents',
  },
  {
    id: 'ai',
    title: 'AI Assistant',
    description: 'Customize answers, preferences and smart suggestions.',
    articlesCount: 5,
    readTime: '2 min read',
    category: 'ai',
  },
  {
    id: 'security',
    title: 'Security',
    description: 'Manage sessions, QR login and account protection.',
    articlesCount: 4,
    readTime: '3 min read',
    category: 'security',
  },
];

const FALLBACK_METRICS: HelpMetrics = {
  averageResponse: '< 5 min',
  resolutionRate: '98.7%',
  ticketsSolved: 12000,
  satisfaction: '4.9/5',
  supportOnline: true,
};

const FALLBACK_CONTACT_OPTIONS: HelpContactOption[] = [
  {
    id: 'ai-chat',
    title: 'AI Chat',
    description: 'Instant AI assistance for tracking and documents.',
    status: 'available',
    cta: 'Start',
    action: 'chat',
  },
  {
    id: 'live-support',
    title: 'Live Support',
    description: 'Connect with our FedEx AI support team.',
    status: 'available',
    cta: 'Contact',
    action: 'contact',
  },
  {
    id: 'email-support',
    title: 'Email Support',
    description: 'Send a detailed request and receive a written response.',
    status: 'available',
    cta: 'Send',
    action: 'contact',
  },
  {
    id: 'knowledge-base',
    title: 'Knowledge Base',
    description: 'Browse articles, guides and frequently asked questions.',
    status: 'available',
    cta: 'Browse',
    action: 'faq',
  },
];

@Injectable({ providedIn: 'root' })
export class HelpCenterService {
  constructor(private readonly http: HttpClient) {}

  getResources(lang?: string): Observable<HelpResource[]> {
    const q = lang ? `?lang=${encodeURIComponent(lang)}` : '';
    return this.http.get<HelpResource[]>(`${API_BASE_URL}/api/help/resources${q}`);
  }

  getMetrics(): Observable<HelpMetrics> {
    return this.http.get<HelpMetrics>(`${API_BASE_URL}/api/help/metrics`);
  }

  getContactOptions(): Observable<HelpContactOption[]> {
    return this.http.get<HelpContactOption[]>(`${API_BASE_URL}/api/help/contact-options`);
  }

  getArticles(lang?: string): Observable<HelpArticleSummary[]> {
    const q = lang ? `?lang=${encodeURIComponent(lang)}` : '';
    return this.http.get<HelpArticleSummary[]>(`${API_BASE_URL}/api/help/articles${q}`).pipe(
      catchError(() => of(FALLBACK_ARTICLE_SUMMARIES.map((a) => ({ ...a })))),
    );
  }

  getArticle(slug: string, lang?: string): Observable<HelpArticle> {
    const q = lang ? `?lang=${encodeURIComponent(lang)}` : '';
    return this.http.get<HelpArticle>(`${API_BASE_URL}/api/help/articles/${encodeURIComponent(slug)}${q}`).pipe(
      catchError(() => {
        const fb = fallbackArticle(slug);
        return fb ? of(fb) : of(null as unknown as HelpArticle);
      }),
    );
  }

  createTicket(payload: CreateTicketPayload): Observable<CreateTicketResponse> {
    return this.http.post<CreateTicketResponse>(`${API_BASE_URL}/api/support/tickets`, payload);
  }

  getMyTickets(): Observable<SupportTicketListResponse> {
    return this.http.get<SupportTicketListResponse>(`${API_BASE_URL}/api/support/tickets/me`);
  }

  getTicket(id: number): Observable<unknown> {
    return this.http.get(`${API_BASE_URL}/api/support/tickets/${id}`);
  }

  fallbackResources(): HelpResource[] {
    return FALLBACK_RESOURCES.map((r) => ({ ...r }));
  }

  fallbackMetrics(): HelpMetrics {
    return { ...FALLBACK_METRICS };
  }

  fallbackContactOptions(): HelpContactOption[] {
    return FALLBACK_CONTACT_OPTIONS.map((o) => ({ ...o }));
  }

  fallbackArticles(): HelpArticleSummary[] {
    return FALLBACK_ARTICLE_SUMMARIES.map((a) => ({ ...a }));
  }
}

File: fedex-main/globex-fedex-main/backend/app/services/help_center_service.py
from __future__ import annotations

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models.support_ticket import SupportTicket, SupportTicketStatus
from app.services.faq_service import load_faq

_CATEGORY_PATTERNS: dict[str, tuple[str, ...]] = {
    "tracking": ("track", "suivi", "colis", "parcel", "delivery", "shipment", "eta", "livraison"),
    "documents": ("document", "pdf", "excel", "export", "proof", "pod", "report", "archive"),
    "ai": ("ai", "assistant", "claude", "chat", "preference", "instruction", "smart"),
    "security": ("security", "password", "session", "qr", "login", "account", "sécurité"),
}

_RESOURCES = [
    {
        "id": "tracking",
        "title": "Tracking issues",
        "description": "Track parcels, ETA, exceptions and delivery problems.",
        "readTime": "2 min read",
        "category": "tracking",
    },
    {
        "id": "documents",
        "title": "Documents & POD",
        "description": "Manage proofs of delivery, exports and archived reports.",
        "readTime": "3 min read",
        "category": "documents",
    },
    {
        "id": "ai",
        "title": "AI Assistant",
        "description": "Customize answers, preferences and smart suggestions.",
        "readTime": "2 min read",
        "category": "ai",
    },
    {
        "id": "security",
        "title": "Security",
        "description": "Manage sessions, QR login and account protection.",
        "readTime": "3 min read",
        "category": "security",
    },
]

_CONTACT_OPTIONS = [
    {
        "id": "ai-chat",
        "title": "AI Chat",
        "description": "Instant AI assistance for tracking and documents.",
        "status": "available",
        "cta": "Start",
        "action": "chat",
    },
    {
        "id": "live-support",
        "title": "Live Support",
        "description": "Connect with our FedEx AI support team.",
        "status": "available",
        "cta": "Contact",
        "action": "contact",
    },
    {
        "id": "email-support",
        "title": "Email Support",
        "description": "Send a detailed request and receive a written response.",
        "status": "available",
        "cta": "Send",
        "action": "contact",
    },
    {
        "id": "knowledge-base",
        "title": "Knowledge Base",
        "description": "Browse articles, guides and frequently asked questions.",
        "status": "available",
        "cta": "Browse",
        "action": "faq",
    },
]

_FALLBACK_COUNTS = {"tracking": 8, "documents": 6, "ai": 5, "security": 4}


def _faq_matches_category(text: str, category: str) -> bool:
    blob = text.lower()
    patterns = _CATEGORY_PATTERNS.get(category, ())
    return any(p in blob for p in patterns)


def get_help_resources(lang: str | None = None) -> list[dict]:
    items, _ = load_faq(lang)
    resources: list[dict] = []
    for base in _RESOURCES:
        cat = base["category"]
        matched = [
            item
            for item in items
            if _faq_matches_category(f"{item.question} {item.answer}", cat)
        ]
        count = len(matched) if matched else _FALLBACK_COUNTS.get(cat, 4)
        resources.append({**base, "articlesCount": count})
    return resources


def get_help_metrics(db: Session) -> dict:
    total = int(db.scalar(select(func.count()).select_from(SupportTicket)) or 0)
    resolved = int(
        db.scalar(
            select(func.count())
            .select_from(SupportTicket)
            .where(SupportTicket.status.in_([SupportTicketStatus.resolved, SupportTicketStatus.closed]))
        )
        or 0
    )
    tickets_solved = resolved if resolved > 0 else 12000
    resolution_rate = f"{(resolved / total * 100):.1f}%" if total > 0 else "98.7%"
    return {
        "averageResponse": "< 5 min",
        "resolutionRate": resolution_rate,
        "ticketsSolved": tickets_solved,
        "satisfaction": "4.9/5",
        "supportOnline": True,
    }


def get_contact_options() -> list[dict]:
    return [dict(opt) for opt in _CONTACT_OPTIONS]


def generate_ticket_number(ticket_id: int) -> str:
    from datetime import datetime, timezone

    year = datetime.now(timezone.utc).year
    return f"SUP-{year}-{ticket_id:04d}"
